### ISSUE TITLE
Seemingly unnecessary type annotation of instance variables

### ISSUE THREAD

**Issue by jaredbeck** [sorbet/sorbet#1666]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0Aclass%20Banana%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7B%20params(color%3A%20String).void%20%7D%0A%20%20def%20initialize(color)%0A%20%20%20%20%40color%20%3D%20color%0A%0A%20%20%20%20%23%20The%20error%206002%20could%20be%20fixed%20by%0A%20%20%20%20%23%20repeating%20the%20type%20information%20already%0A%20%20%20%20%23%20given%20in%20the%20sig.%0A%20%20%20%20%23%20%40color%20%3D%20T.let(color%2C%20String)%0A%20%20end%0Aend%0A)

```ruby
# typed: strict
class Banana
  extend T::Sig

  sig { params(color: String).void }
  def initialize(color)
    @color = color

    # The error 6002 could be fixed by
    # repeating the type information already
    # given in the sig.
    # @color = T.let(color, String)
  end
end
```

#### Observed output

```
editor.rb:7: Use of undeclared variable @color https://srb.help/6002
     7 |    @color = color
            ^^^^^^
Errors: 1
```

#### Expected behavior

No error should be produced.

#### Analysis

The type of `@color` can be inferred from the method signature.

I don't know enough to say if the current behavior (requiring the type information to be repeated in a `T.let`) is necessary, but if so, the repetition is unfortunate.

#### :heart:

sorbet is my favorite thing about ruby in 2019. Keep up the good work!

**Comment by jez** [comment]
Thanks for the report. Sorbet has opted to make declaring a class's API (internal and external). Only local variables are inferred—this is one of the key ways Sorbet achieves really fast type checking speeds, because type checking one method becomes completely independent of all other methods and can be done in parallel.

Some time ago I added an FAQ section about this:

https://sorbet.org/docs/type-annotations#why-do-i-need-to-repeat-types-from-the-constructor

As you mention, this is unfortunate, and we might change this in the future, but we have no immediate plans to.

**Comment by jaredbeck** [comment]
Thanks Jake. Sorry I missed the docs; nice write-up.

> As you mention, this is unfortunate ..

Thinking about this a bit more, java has the same limitation, so we're in good company.

```java
class Banana {
  public int mass;
  public Banana(int mass) {
    this.mass = mass;
  }
```

Oddly, the exact same level of duplicate-type-annotation never bothered me in java.

Feel free to close this, or maybe keep it open if you think it will keep other people from raising the same issue.

**Comment by connorshea** [comment]
I think this is resolved by #2230?
